tikv/deadlock.go

```diff
@@ -0,0 +1,217 @@
+package tikv
+
+import (
+	"context"
+	"google.golang.org/grpc"
+	"sync/atomic"
+	"time"
+
+	"github.com/ngaut/log"
+	"github.com/ngaut/unistore/pd"
+	"github.com/ngaut/unistore/util/lockwaiter"
+	deadlockPb "github.com/pingcap/kvproto/pkg/deadlock"
+	"github.com/pingcap/kvproto/pkg/metapb"
+)
+
+// Follower will send detection rpc to Leader
+const (
+	Follower = iota
+	Leader
+)
+
+// DeadlockDetector is a util used for distributed deadlock detection
+type DeadlockDetector struct {
+	detector  *Detector
+	pdClient  pd.Client
+	storeMeta *metapb.Store
+	sendCh    chan *deadlockPb.DeadlockRequest
+	waitMgr   *lockwaiter.Manager
+	streamCli deadlockPb.Deadlock_DetectClient
+
+	// these fields used by multiple thread
+	role int32
+}
+
+// refreshFirstRegionLeader will send request to pd to find out the
+// current leader node for the first region
+func (dt *DeadlockDetector) refreshFirstRegionLeader() (string, error) {
+	// find first region from pd, get the first region leader
+	ctx := context.Background()
+	_, leaderPeer, err := dt.pdClient.GetRegion(ctx, []byte{})
+	if err != nil {
+		log.Errorf("get first region failed, err: %v", err)
+		return "", err
+	}
+	leaderStoreMeta, err := dt.pdClient.GetStore(ctx, leaderPeer.GetStoreId())
+	if err != nil {
+		log.Errorf("get store=%d failed, err=%v", leaderPeer.GetStoreId(), err)
+		return "", err
+	}
+	log.Warnf("refreshFirstRegionLeader leader_peer=%v addr=%s", leaderPeer, leaderStoreMeta.GetAddress())
+	return leaderStoreMeta.GetAddress(), nil
+}
+
+// rebuildStreamClient builds connection to the first region leader,
+// it's not thread safe and should be called only by `DeadlockDetector.Start` or `DeadlockDetector.SendReqLoop`
+func (dt *DeadlockDetector) rebuildStreamClient() (string, error) {
```

Comment:
Return the address looks weird.
We can log the address here and don't return the address.

Reply:
return the streamClient better fits the method name, we can start the recvLoop goroutine in the caller code